server/email_verification.go

```diff
@@ -190,52 +193,110 @@ type emailVerifiedTemplateData struct {
 	Message string
 }
 
-func handleEmailVerifyFunc(verifiedTpl *template.Template, issuer url.URL, keysFunc func() ([]key.PublicKey,
+func handleEmailVerifyFunc(verifiedTpl *template.Template, s *sessionmanager.SessionManager, issuer url.URL, keysFunc func() ([]key.PublicKey,
 	error), userManager *manager.UserManager) http.HandlerFunc {
 
 	return func(w http.ResponseWriter, r *http.Request) {
 		q := r.URL.Query()
 		token := q.Get("token")
 
+		runtpl := func(error string, message string, statusCode int) {
+			execTemplateWithStatus(w, verifiedTpl, emailVerifiedTemplateData{
+				Error:   error,
+				Message: message,
+			}, statusCode)
+		}
+
 		keys, err := keysFunc()
 		if err != nil {
-			execTemplateWithStatus(w, verifiedTpl, emailVerifiedTemplateData{
-				Error:   "There's been an error processing your request.",
-				Message: "Please try again later.",
-			}, http.StatusInternalServerError)
+			runtpl("There's been an error processing your request.", "Please try again later.", http.StatusInternalServerError)
 			return
 		}
 
 		ev, err := user.ParseAndVerifyEmailVerificationToken(token, issuer, keys)
 		if err != nil {
-			execTemplateWithStatus(w, verifiedTpl, emailVerifiedTemplateData{
-				Error:   "Bad Email Verification Token",
-				Message: "That was not a verifiable token.",
-			}, http.StatusBadRequest)
+			runtpl("Bad Email Verification Token", "That was not a verifiable token.", http.StatusBadRequest)
+			return
+		}
+
+		aud, ok, err := ev.Claims.StringClaim("aud")
+		if !ok {
+			runtpl("Unable to get 'aud' from token", "That was not a verifiable token.", http.StatusBadRequest)
+			return
+		}
+
+		sub, ok, err := ev.Claims.StringClaim("sub")
+		if !ok {
+			runtpl("Unable to get 'sub' from token", "That was not a verifiable token.", http.StatusBadRequest)
+			return
+		}
+
+		callback, ok, err := ev.Claims.StringClaim("http://coreos.com/email/verification-callback")
+		if !ok {
+			runtpl("Unable to get callback URL from token", "That was not a verifiable token.", http.StatusBadRequest)
+			return
+		}
```

Comment:
indeed, but we need to extract these params here too using `StringClaim`, because `ParseAndVerifyEmailVerificationToken` func does not return them, so I think it's ok if we check errors here again. what do you think? do you prefer returning these params in the `ParseAndVerifyEmailVerificationToken` func?